### PR TITLE
SelectAllTextBehavior

### LINKED ISSUE #82: [Proposal] SelectAllTextEffect
# SelectAllTextEffect

* [x] Proposed
* [x] Prototype
* [x] Implementation
  * [x] iOS Support
  * [x] Android Support
  * [x] macOS Support
  * [x] Windows Support
* [ ] Unit Tests: Not Started
* [x] Sample
* [ ] Documentation: Not Started

## Summary
[summary]: #summary

Highlights all text when an InputView is selected

## Detailed Design
[design]: #detailed-design

### SelectAllTextEffect.shared.cs

```cs
public class SelectAllTextEffect : RoutingEffect
{
}
```

### SelectAllTextEffect.android.cs

```cs
public class SelectAllTextEffect : PlatformEffect
{
  EditText EditText => (EditText)Control;
  
  protected override void OnAttached() => EditText?.SetSelectAllOnFocus(true);
  
  protected override void OnDetached() => EditText?.SetSelectAllOnFocus(false);
}
```

### SelectAllTextEffect.ios.cs

```cs
public class SelectAllTextEffect : PlatformEffect
{
  protected override void OnAttached() => ApplyEffect(true);
  
  protected override void OnDetached() => ApplyEffect(false);
  
  void ApplyEffect(bool apply) => ApplyToControl(Control, apply);

  bool ApplyToControl<T>(T controlType, bool apply) => controlType switch
  {
	UITextField textField => ApplyToUITextField(textField, apply),
	UITextView => ApplyToUITextView(apply),
	_ => throw new NotSupportedException($"Control of type: {controlType?.GetType()?.Name} is not supported by this effect.")
  };
}
```

## Usage Syntax
[usage]: #usage-syntax

<!-- Here is an example from CharactersValidationBehavior:

### XAML Usage

```xml
<Entry
    <Entry.Effects>
        <local: SelectAllTextEffect/>
    </Entry.Effects>
</Entry>

```

### C# Usage

```cs
var entry = new Entry();
entry.Effects.Add(new SelectAllTextEffect());
```

### PR DESCRIPTION
**Pull request by pictos**
### Description of Change ###

Since MAUI is GA 🎊 I presume that our code freeze is ending. With that here's the first `Effect` converted to `Behavior` using the new `Behavior API`.
 ### Linked Issues ###
 

 - Fixes #82

 ### PR Checklist ###
 
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/XYZ 

 ### Additional information ###
I'm looking at how to implement the unit test since I need some platform code to validate if the text is selected. Any idea is welcome.

### CODE REVIEW

File: src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.ios.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+		return true;
+	}
+
+	void OnEditingDidBegin(object? sender, EventArgs e)
+	{
+		if (sender is not UITextField textfield)
+		{
+			return;
+		}
```
Comment (brminnick) [inline src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.ios.cs:61]: Should this throw an `ArgumentException` or an `InvalidOperationException`?


Is any other type subscribing to this event handler that isn't a `UITextField`? If not, let's throw an Exception.


Or, better, just cast it:

```suggestion
		var texfield = (UITextField)sender;
```

File: samples/CommunityToolkit.Maui.Sample/Pages/Behaviors/SelectAllTextBehaviorPage.xaml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+							<mct:SelectAllTextBehavior />
+						</Editor.Behaviors>
+						<Editor.Text>
+							Yeah, but your scientists were so preoccupied with whether or not they could, they didn't stop to think if they should. Must go faster... go, go, go, go, go! You know what? It is beets. I've crashed into a beet truck. Is this my espresso machine? Wh-what is-h-how did you get my espresso machine?
+							Just my luck, no ice. I gave it a cold? I gave it a virus. A computer virus. Must go faster... go, go, go, go, go! God creates dinosaurs. God destroys dinosaurs. God creates Man. Man destroys God. Man creates Dinosaurs. Yeah, but your scientists were so preoccupied with whether or not they could, they didn't stop to think if they should.
+						</Editor.Text>
+					</Editor>
+				</Frame>
+			</VerticalStackLayout>
+		</VerticalStackLayout>
```
Comment (brminnick) [inline samples/CommunityToolkit.Maui.Sample/Pages/Behaviors/SelectAllTextBehaviorPage.xaml:83]: Let's avoid nesting multiple VerticalStackLayouts inside of a VerticalStackLayout. 

Let's either use a `Grid`, or a `FlexLayout`, or edit each control's `Padding` + `Margin` + `Spacing` to fine-tune the groupings. 

https://docs.microsoft.com/en-us/xamarin/xamarin-forms/deploy-test/performance#choose-the-correct-layout
> ... don't attempt to reproduce the appearance of a specific layout by using combinations of other layouts, as this results in unnecessary layout calculations being performed. For example, don't attempt to reproduce a Grid layout by using a combination of StackLayout instances.

**Review by brminnick — COMMENTED** [review]
Thanks Pedro! Exciting to start working on v1.1.0 now!!

Here's a couple comments after a quick glance at the code. I'll spin it up on my machine and play around with it tomorrow 🙌

File: src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.ios.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+			_ => throw new NotSupportedException($"Control of type: {platformView?.GetType()?.Name} is not supported by this effect.")
+		};
+	}
+
+	bool ApplyToUITextField(UITextField textField, bool apply)
+	{
+		if (textField is null)
+		{
+			return false;
+		}
```
[the remark this answers is not in the document]
Comment (pictos) [reply src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.ios.cs:42]: From XCT we didn't throw the exception, so not sure if we should change the behavior or not

File: src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.ios.cs (hunk 1 above)
Comment (pictos) [reply src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.ios.cs:61]: same as before, I just maintained the previous behavior, as you can see [here](https://github.com/xamarin/XamarinCommunityToolkit/blob/22b1d4ccf45555c2ed89ecb94e9ff93a5cf61ab3/src/CommunityToolkit/Xamarin.CommunityToolkit/Effects/SelectAllText/SelectAllTextEffect.ios.cs#L73-L74). Do you think that we should throw an exception?

If I'm not wrong, throwing an exception at this level could crash the app, not sure if the behavior is too important to the point of being able to break the application

File: src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.ios.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (brminnick) [reply src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.ios.cs:42]: We don't need to adhere to the `Xamarin.CommunityToolkit` implementation. In fact, I highly recommend we don't for every feature we add to `CommunityToolkit.Maui`!

I always advocate for throwing Exceptions over printing an error message to the console for two reasons:
1. They immediately help the developer locate the problem in their code
2. If the method fails without throwing an exception, the developer will experience an unexpected behavior from our library 
    * This leads to them wasting time debugging it, looking at our source code, etc. trying to figure out the problem
    * The developer may open an Issue on our repo, leading to wasted time for us explaining the expected implementation and behavior to them

[the remark this answers is not in the document]
Comment (pictos) [reply src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.ios.cs:42]: @brminnick  I'm fine with that, I'll update the code to throw exceptions

File: src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.ios.cs (hunk 1 above)
Comment (pictos) [reply src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.ios.cs:61]: Resolved in another discussion

**Review by brminnick — CHANGES_REQUESTED** [review]
Hey Pedro! 

I found a bug when selecting the **Editor** text on macOS; when I tap on the Editor, it does not select all of the text. 

It works fine on iOS though, which is odd.

Here's a screen recording demoing iOS (right) and macOS (left):

ScreenFlow

**Comment by pictos** [comment]
Thanks, @brminnick, maybe it's some API detail. I'll set up my mac environment and try it here on my end.

**Comment by pictos** [comment]
@brminnick I ran the sample using the VSCode, and I would like to confirm that the environment is not deceiving me. Do you mind running our sample and putting a breakpoint [here](https://github.com/CommunityToolkit/Maui/blob/8a6b34725d7e2bf5b6feda1030c45f7d98344794/src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.ios.cs#L83) and [here](https://github.com/CommunityToolkit/Maui/blob/8a6b34725d7e2bf5b6feda1030c45f7d98344794/src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.ios.cs#L95).

On my test the `Focused` event isn't trigger on Editor.

**Comment by pictos** [comment]
@brminnick FYI I opened an issue report on `Xamarin-macios` repo.
- https://github.com/xamarin/xamarin-macios/issues/15156

**Comment by brminnick** [comment]
Thanks @pictos! 

I just saw Rolf's response to https://github.com/xamarin/xamarin-macios/issues/15156.

What do you think? Should we add a work-around for macOS that modifies this default behavior?

**Comment by pictos** [comment]
@brminnick I would say that we can just mark it as not supported on mac-catalyst and I'll do some investigation in the future. I didn't find anything on the internet about how to workaround this.

This one was supposed to be easy peasy 😄

**Comment by pictos** [comment]
@brminnick the `NotSupportedOSAttribute` should be used on a `public` method, but we don't have it. I thought to add it inside the `SelectAllTextBehavior` class, but it works well for `Entry` control. Do you have an idea where we can add it?

**Comment by brminnick** [comment]
Dang - good point. 

You're right - if we put the attribute on the class (like this), then users will assume that it doesn't work at all on macOS, which isn't accurate because it _does_ work for `Entry`:

```cs
[UnsupportedOSPlatform("maccatalyst")]
public class SelectAllTextBehavior : PlatformBehavior<InputView>
```

Let's do this:
1. Open an Issue in this repo documenting the behavior for Editor + macOS
2. Update the comment and the `Console.Writeline` to point to this known issue
3. Ship it 🚀

File: src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.macios.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+		textfield.PerformSelector(new Selector("selectAll"), null, 0.0f);
+	}
+
+	// MacCatalyst support blocked: https://github.com/xamarin/xamarin-macios/issues/15156
+	bool ApplyToUITextView(bool shouldApply)
+	{
+		if (OperatingSystem.IsMacCatalyst())
+		{
+			Console.WriteLine("MacCatalyst doesn't support this feature on the editor. for more info see https://github.com/CommunityToolkit/Maui/issues/432");
```
Comment (brminnick) [inline src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.macios.cs:77]:
```suggestion
			Console.WriteLine("WARNING: `SelectAllTextBehavior` does not support `Microsoft.Maui.Controls.Editor` on MacCatalyst. For more information, see https://github.com/CommunityToolkit/Maui/issues/432");
```

Comment (brminnick) [reply src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/SelectAllText/SelectAllTextBehavior.macios.cs:77]: Just a quick suggestion to include `SelectAllTextBehavior` in the message and add WARNING to help it stand out in the logs

**Review by brminnick — APPROVED** [review]

**Review by brminnick — CHANGES_REQUESTED** [review]
Apologies for approving than requesting changes. 

I just noticed we're missing unit tests. Let's add those and submit a Docs PR before merging this PR 👍

**Comment by pictos** [comment]
@brminnick I tried to create the unit test and it couldn't be done... Because the platformBehavior uses the native APIs, if you see the code there's no `.shared` file, so it doesn't target .NET6. I'll move the device_test forward to make sure that we can test everything.



image

**Pull request merged**